### DISCUSSION TITLE
Modifications for Density Based Topology Optimisation

### DISCUSSION THREAD

**Discussion by jjahammond** [mdolab/dafoam#80] [General]
Hi Ping,

I'm a PhD student and so far my research has focused on turbulence modelling in density based topology optimisation using continuous adjoints in OpenFoam. I am interested in trying a discrete adjoint approach as our current continuous adjoint implementations still rely on a constant eddy viscosity assumption.

That said I'm interested in trying to modify DAFoam to run density based topology optimisations (e.g. by including a resistance term `fvm::Sp(alpha,U)` into the UEqnSimple.H of DASimpleFoam and iteratively updating `alpha` based upon the sensitivity `dC/dalpha`). I wonder how tractable you think this would be given the current focus of the DAFoam library is for shape optimisation? 

Thanks and I look forward to hearing from you,
James

**Comment by friedenhe** [comment]
We recently added the capability to compute dF/dBeta, where beta is a scalar field in the turbulence equation. We are about to release a new version 2.2.1 that will allow computing dF/dAlpha, where alpha is a source term (alpha*U) added to the U equation (This happens to be exactly what you need). So stay tuned.

**Reply by jjahammond** [reply]
That's really exciting news! Do you have some idea of the timeframe for this release?

I have also been having issues compiling everything from source. I run Ubuntu 20.04 LTS (and use OpenFOAM 7 from the OpenFOAM Foundation rather than the ESI-OpenCFD releases). Would you recommend following your exact dependencies or should it be possible to setup on Ubuntu 20.04?

**Reply by friedenhe** [reply]
It will happen sometime this month. Note that we will only add the capability to compute dF/dAlpha, you still need to develop a runScript.py and OpenFOAM configuration to do topology optimization. DAFoam only supports the OpenFOAM branch from ESI, it does not support OpenFOAM 7. We haven't tested the installation on Ubuntu 20.04 yet.

**Reply by jjahammond** [reply]
Okay perfect. Thanks for your help - this has been incredibly useful. I will keep an eye out for the release and also let you know if I get any further with setup on Ubuntu 20.04.

**Reply by friedenhe** [reply]
Try this case with DAFoam v2.2.1.
[NACA0012_AlphaPorosity.tar.gz](https://github.com/mdolab/dafoam/files/5827386/NACA0012_AlphaPorosity.tar.gz)

**Reply by jjahammond** [reply]
Thanks a lot this is really exciting, I will give it a go! What's the current optimisation set up for this case? i.e. If not already set up as a topology optimisation what does it do with the gradient information w.r.t alphaPorosity?

**Reply by friedenhe** [reply]
It is an optimization setup. It will minimize the drag by changing the alpha field. I suggest you first go through the Get Started section on https://dafoam.github.io for a shape optimization case and understand the run scripts. Then, you can adjust the alpha porosity case with your own geometry and boundary conditions.

**Comment by schewef** [comment]
Dear friedenhe,

first of all thank you very much for providing this great framework!

I am also very interested in testing the possibilities of DAFoam regarding density based topology optimization.
I noticed that the current implementation is very dependent on pyGeo, which as I understand is for shape variation. 
Is it possible to set up an optimization without using the pyGeo module, or is it possible to define design variables in pyGeo like the porosity that have nothing to do with shape optimization at all? I already made some tries to get rid of pyGeo but failed.

If it is not possible yet, do you plan to provide such functionality in the future?
I would also be happy with an pure calculation of sensitivities towards the porosity. If I got you right in the discussion above that is already available, even by the python API? 

Thank you very much in advance!

**Reply by friedenhe** [reply]
@schewef Yes, the computation of porosity derivative is available since v2.2.2. Unfortunately, we have to set up a dummy pyGeo object to get DAFoam running. You need to generate a large FFD box in genFFD.py to cover all the walls and set up designSurface in runScript.py, although the design surface does not change. In addition, you have to add a reference axis. In the future, we will get rid of the dependency for pyGeo for topo optimization problems. But this is not our priority at this moment and will take a while to happen.

**Reply by schewef** [reply]
Thank you for the feedback! 
I will try to calculate the sensitivities of a simple topology optimization problem using your suggestions, to check if the accuracy fits the needs of the problem I am working on.
If so I can imagine to suggest some code adaptions to improve the usability for topology optimization problems.

**Reply by friedenhe** [reply]
Great. We are in the process of putting together a tutorial for density-based topology optimization. Let us know if you have any suggestions for the codes, or you may want to submit a pull request through Github. Stay tuned.

**Reply by schewef** [reply]
Dear @friedenhe
i did some first studies and so far I am quite sufficient with the results! I also had a first look into the code. Am I correct in assuming that no kind of density or sensitivity filtering has been implemented so far?

**Reply by friedenhe** [reply]
That is right, there is no filtering for alpha in the current version.

**Comment by friedenhe** [comment]
@jjahammond @schewef Check [this new tutorial](https://dafoam.github.io/mydoc_tutorials_topo_pitzdaily.html) for density-based topology optimization.

**Reply by schewef** [reply]
Wow, thank you very much! This really takes me further!

**Reply by zzhya** [reply]
I don't have SNOPT, can I use IPOPT to optimize?

**Reply by zzhya** [reply]
Hello, I don't have SNOPT, can I use IPOPT to do fluid Topology Optimisation in DAFOAM(density-based topology optimization case for the pitzDaily case in OpenFOAM)?

**Reply by friedenhe** [reply]
We tested IPOPT, and it did not work with its default setup. But if you are familiar with IPOPT and can tweak its parameters to make it work, please let us know.

**Reply by zzhya** [reply]
IPOPT and SNOPT have the same functionality and I think IPOPT can be successful, I will give it a try. BTW, do you have a detailed result file with SNOPT topology optimization? Can you send me a copy if you have? Here's my email: chunniunai0.0@gmail.com. If I have success with IPOPT optimization, I can compare them.

**Reply by ludocampa** [reply]
Hi @zzhya, did you managed the setup with IPOPT?

**Comment by NeuroNexusAI** [comment]
Modifications for density-based topology optimization include adaptive mesh refinement, sensitivity filtering, and regularization. These enhance results, reduce issues, and improve manufacturability. Adding multi-material or dynamic load considerations can further optimize designs.
